framework/src/outputs/TopResidualDebugOutput.C

```diff
@@ -71,8 +73,25 @@ TopResidualDebugOutput::printTopResiduals(const NumericVector<Number> & residual
       {
         const auto & subdomain_ids = mesh.getNodeBlockIds(*node);
         dof_id_type dof_idx = node->dof_number(_sys.number(), var, 0);
-        vec[j] =
-            TopResidualDebugOutputTopResidualData(var, subdomain_ids, nd, *node, residual(dof_idx));
+        vec[j] = TopResidualDebugOutputTopResidualData(
+            var, subdomain_ids, nd, *node, residual(dof_idx), false, true);
+        j++;
+      }
+  }
+
+  // Loop over all elemental variables
+  for (const auto & elem : as_range(mesh.activeLocalElementsBegin(), mesh.activeLocalElementsEnd()))
+  {
+    dof_id_type elem_id = elem->id();
+    const SubdomainID subdomain_id = elem->subdomain_id();
+
+    for (unsigned int var = 0; var < elem->n_vars(_sys.number()); ++var)
+      if (elem->n_dofs(_sys.number(), var) >
+          0) // this check filters scalar variables (which are clearly not a dof on every node)
```

Comment:
This is for subdomain restricted variables. For scalar variable DOFs there is `DofMap::n_SCALAR_dofs`